src/dawn/IIR/FieldAccessMetadata.h

```diff
@@ -26,38 +28,122 @@ namespace iir {
   }
 
-  bool isAccessIDAVersion(const int accessID) { return versionIDs_.count(accessID); }
+  void removeID(const int accessID) {
+    if(derivedInfo_.versionIDs_.count(accessID) > 0) {
+      // Remove the field from the versions of it's original field
+      int originalID = getOriginalVersionOfAccessID(accessID);
+      auto vec = variableVersionsMap_[originalID];
+      std::remove_if(vec->begin(), vec->end(), [&accessID](int id) { return id == accessID; });
+      // Remove the backward map to it's origin
+      derivedInfo_.versionToOriginalVersionMap_.erase(accessID);
+      // and clear it from the list of all versioned fields
+      derivedInfo_.versionIDs_.erase(accessID);
+    } else if(variableVersionsMap_.count(accessID) > 0) {
+      int newOriginalID = variableVersionsMap_[accessID]->at(0);
+      // This is now an original field
+      derivedInfo_.versionIDs_.erase(newOriginalID);
+      // change the orignial to version map:
+      std::shared_ptr<std::vector<int>> newVec = std::make_shared<std::vector<int>>();
+      for(auto ID : (*variableVersionsMap_[accessID])) {
+        if(ID != newOriginalID) {
+          newVec->push_back(ID);
+          // And hook the versions to it's new original
+          derivedInfo_.versionToOriginalVersionMap_.erase(ID);
+          derivedInfo_.versionToOriginalVersionMap_.emplace(ID, newOriginalID);
+        }
+      }
+      variableVersionsMap_.erase(accessID);
+      variableVersionsMap_.emplace(newOriginalID, newVec);
+    } else {
+      DAWN_ASSERT_MSG(false, "field is not versioned");
+    }
+  }
+
+  bool isAccessIDAVersion(const int accessID) const {
+    return derivedInfo_.versionIDs_.count(accessID);
+  }
 
   int getOriginalVersionOfAccessID(const int accessID) const {
-    return versionToOriginalVersionMap_.at(accessID);
+    if(isAccessIDAVersion(accessID)) {
+      return derivedInfo_.versionToOriginalVersionMap_.at(accessID);
+    } else {
+      DAWN_ASSERT_MSG(0, "try to access original version of non-versioned field");
+    }
+    return 0;
   }
-  const std::unordered_set<int>& getVersionIDs() const { return versionIDs_; }
 
-  VariableVersions() = default;
+  const std::unordered_set<int>& getVersionIDs() const { return derivedInfo_.versionIDs_; }
+
+  const std::unordered_map<int, std::shared_ptr<std::vector<int>>>& getvariableVersionsMap() const {
+    return variableVersionsMap_;
+  }
 
   json::json jsonDump() const;
+
+  // now that we only have getters and setters this should always be in a consistent state.
+  // Should we remove this enirely, have it here as a safety valve if one wants to mess with the
+  // versioning and needs a clean update or have this as part of the update-level step for the
+  // top-most level holding the metadata?
+  // I personally think option 1 should be desired as this is (hopefully) never necessary but I've
+  // put it here if the reviewers think this would be helpful
+  void recomputeDerivedInfo() {
```

Comment:
I think this has now no use, so it should be removed. 
As a side note, having derived data structures makes me think that maybe a Design by Contract framework, where things like this are checked in an invariant expression, could be helpful, but it may be too much of a long step.